lib/Solver/Solver.cpp

```diff
@@ -0,0 +1,363 @@
+  }
+  switch (R.getKind()) {
+  case feature::Relationship::RelationshipKind::RK_ALTERNATIVE:
+    Solver->add(z3::implies(ParentOption, z3::atmost(V, 1) & z3::mk_or(V)));
+    break;
+  case feature::Relationship::RelationshipKind::RK_OR:
+    Solver->add(z3::implies(ParentOption, z3::mk_or(V)));
+    break;
+  }
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::addConstraint(feature::Constraint &ConstraintToAdd) {
+  SolverConstraintVisitor SCV(this);
+  bool Succ = SCV.addConstraint(&ConstraintToAdd);
+  if (!Succ) {
+    return SolverErrorCode::NOT_SUPPORTED;
+  }
+  return Ok();
+}
+
+Result<SolverErrorCode, std::unique_ptr<bool>>
+Z3Solver::hasValidConfigurations() {
+  if (Solver->check() == z3::sat) {
+    return Ok(std::make_unique<bool>(true));
+  }
+  return Ok(std::make_unique<bool>(false));
+}
+
+Result<SolverErrorCode, std::unique_ptr<vara::feature::Configuration>>
+Z3Solver::getNextConfiguration() {
+  // Add previous configuration as a constraint
+  excludeCurrentConfiguration();
+
+  // Retrieve the next configuration
+  return getCurrentConfiguration();
+}
+
+Result<SolverErrorCode, std::unique_ptr<uint64_t>>
+Z3Solver::getNumberValidConfigurations() {
+  Solver->push();
+  std::unique_ptr<uint64_t> Count = std::make_unique<uint64_t>(0);
+  while (Solver->check() == z3::sat) {
+    excludeCurrentConfiguration();
+    (*Count)++;
+  }
+  Solver->pop();
+  return Count;
+}
+
+Result<
+    SolverErrorCode,
+    std::unique_ptr<std::vector<std::unique_ptr<vara::feature::Configuration>>>>
+Z3Solver::getAllValidConfigurations() {
+  Solver->push();
+  std::unique_ptr<std::vector<std::unique_ptr<vara::feature::Configuration>>>
+      Vector(new std::vector<std::unique_ptr<vara::feature::Configuration>>);
+  while (Solver->check() == z3::sat) {
+    auto Config = getCurrentConfiguration().extractValue();
+    Vector->insert(Vector->begin(), std::move(Config));
+    excludeCurrentConfiguration();
+  }
+  Solver->pop();
+  return Vector;
+}
+
+Result<SolverErrorCode>
+Z3Solver::setBinaryFeatureConstraints(const feature::BinaryFeature &Feature,
+                                      bool IsInAlternativeGroup) {
+  // Add constraint to parent
+  Solver->add(z3::implies(
+      *OptionToVariableMapping[Feature.getName()],
+      *OptionToVariableMapping[Feature.getParentFeature()->getName()]));
+
+  if (!IsInAlternativeGroup && !Feature.isOptional()) {
+    Solver->add(z3::implies(
+        *OptionToVariableMapping[Feature.getParentFeature()->getName()],
+        *OptionToVariableMapping[Feature.getName()]));
+  }
+
+  SolverConstraintVisitor SCV(this);
+
+  return Ok();
+}
+
+Result<SolverErrorCode> Z3Solver::excludeCurrentConfiguration() {
+  if (Solver->check() == z3::unsat) {
+    return UNSAT;
+  }
+  z3::model M = Solver->get_model();
+  z3::expr Expr = Context.bool_val(false);
+  for (auto Option : OptionToVariableMapping.keys()) {
+    z3::expr OptionExpr = *OptionToVariableMapping[Option];
```

Comment:
If you need both, key and value normally iterate over the map using the iterator then you have direct access to both and don't have to lookup the value later on.

Reply:
Thanks; I will do that in the next revision.